### DISCUSSION TITLE
Recommended approach for integrating with *non*-Yew Rust/WASM web app

### DISCUSSION THREAD

**Discussion by michaelhagel** [perspective-dev/perspective#1892] [Q&A]
*Context*: I have a dashboarding application written with [Sycamore](https://github.com/sycamore-rs/sycamore), and I'm leaning towards moving the charting functionality over to Perspective.

*Question*: Is there a recommended way of interfacing with the Perspective Viewer component from a Rust/WASM web app? For simplicity, I first just rendered the Perspective Viewer element via a traditional Sycamore `view!` macro workflow (much like with the `html!` Yew macro), and interacted with the element via an established ID and some `#[wasm_bindgen]` Perspective JS API scripts.

This is very simple and works well enough, but I would prefer to keep as much of the API interaction within Rust itself *where possible*. 

Any recommended approach? Or stick with FFI to Perspective JS API scripts?

**Comment by texodus** [comment]
I'm working on this right now, but there are some unresolved issues.  I expect that when this does work, we won't be making framework-specific wrappers per se (not even Yew), but rather making it easy to call the underlying Rust API directly from a `<perspective-viewer>`;  so basically, in your case, you'll still create this in the `view!` macro as HTML and capture it by ID, but you'll then be able to get a native Rust API directly from this if you link with rust version of the library.

* There's an existing method  [to get the raw pointer to the web component's Rust model](https://github.com/finos/perspective/blob/master/rust/perspective-viewer/src/rust/custom_elements/viewer.rs#L559), which you can then pass back into Rust from JavaScript and cast back to `PerspectiveViewerElement` struct.  You can then call any available public methods on it, and I've recently refactored this API to not cast to `JsValue` where-ever possible (#1846).  However, you must then import `perspective-viewer` as an external crate to link with your app, which we do not currently publish, see (2).
* While the `<perspective-viewer>` web component is written in Rust, it has dependencies on JavaScript libraries like `monaco`, `regular-table`, `d3fc`, etc., some of which are tricky to build.  We'll either need to bundle these in the crate or equivalent somehow, or `wasm-pack`/`trunk` will not work.
* There are also some Typescript wrappers which we'd need to port, and to _fully_ do this (and only use snippets), we'd need to be able to extend an ES6 class in `wasm_bindgen` (in order to call `customElements.define()` method), which I cannot figure out how to do.

I'll post any updates to this as I make progress.

**Reply by michaelhagel** [reply]
Fantastic rundown, thank you for all the details. With respect to (3), I assume this is a bit more difficult than:
```rust
/// Pseudocode here that probably doesn't make sense (sorry)
#[wasm_bindgen]
extern "C" {
    type ES6Class;

    #[wasm_bindgen(extends = ES6Class)]
    type Bar;
}
```
as I believe `wasm-bindgen` has somewhat stalled on providing solutions for extending ES6 classes (reference: https://github.com/rustwasm/wasm-bindgen/issues/210)

I, at least, cannot find anything related to extending ES6 classes. My domain knowledge is limited there, though, so I may be wrong.

I look forward to updates and wish you luck! I'll be on the lookout for open issue tickets related to this as well if I have time.

**Comment by texodus** [comment]
I need the opposite of that (extends a wasm_bindgen generated JavaScript constructor to a Rust struct), precisely what's in the linked issue.  The `extern` version allows you to specify prototype relations for externals, not rust-generated bindings.

There was a previous (now closed) wasm_bindgen PR that was closed which added `js_sys::Function::of()`, which would've returned the necessary constructor functions for a type, https://github.com/rustwasm/wasm-bindgen/pull/1737 .  Without this, there will need to be at least a couple of lines of Javascript necessary to bootstrap the Rust code into the JavaScript Custom Elements API, and while this is not strictly necessary for Rust code itself to run, the other Javascript parts (e.g. the datagrid plugin or expression editor) cannot call rust functions, they communicate back to the Rust code via this API, and thus won't work.

**Reply by michaelhagel** [reply]
Ah, okay, I see my mistake. Looks like I will be continuing with my original solution for now, and that is perfectly fine.

Thank you again for the explanations, and I am excited for progress on both wasm-bindgen and your work here.
